internal/common/testlogger/testlogger.go

```diff
@@ -0,0 +1,148 @@
+		and this cannot easily be patched from the outside.
+
+		so this logger cheats: after a test completes, it logs to stderr rather than TestingT.
+		EVERY ONE of these logs is bad and we should not produce them, but it's causing many
+		otherwise-useful tests to be flaky, and that's a larger interruption than is useful.
+	*/
+	logAfterComplete, err := zap.NewDevelopment()
+	require.NoError(t, err, "could not build a fallback zap logger")
+	replaced := &fallbackTestCore{
+		t:         t,
+		fallback:  logAfterComplete.Core(),
+		testing:   zaptest.NewLogger(t).Core(),
+		completed: &atomic.Bool{},
+	}
+
+	t.Cleanup(replaced.UseFallback) // switch to fallback before ending the test
+
+	return zap.New(replaced)
+}
+
+// NewObserved makes a new test logger that both logs to `t` and collects logged
+// events for asserting in tests.
+func NewObserved(t TestingT) (*zap.Logger, *observer.ObservedLogs) {
+	obsCore, obs := observer.New(zapcore.DebugLevel)
+	z := NewZap(t)
+	z = z.WithOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
+		return zapcore.NewTee(core, obsCore)
+	}))
+	return z, obs
+}
+
+type fallbackTestCore struct {
+	sync.Mutex
+	t         TestingT
+	fallback  zapcore.Core
+	testing   zapcore.Core
+	completed *atomic.Bool
+}
+
+var _ zapcore.Core = (*fallbackTestCore)(nil)
+
+func (f *fallbackTestCore) UseFallback() {
+	f.completed.Store(true)
+}
+
+func (f *fallbackTestCore) Enabled(level zapcore.Level) bool {
+	if f.completed.Load() {
+		return f.fallback.Enabled(level)
+	}
+	return f.testing.Enabled(level)
+}
+
+func (f *fallbackTestCore) With(fields []zapcore.Field) zapcore.Core {
+	// need to copy and defer, else the returned core will be used at an
+	// arbitrarily later point in time, possibly after the test has completed.
+	return &fallbackTestCore{
+		t:         f.t,
+		fallback:  f.fallback.With(fields),
+		testing:   f.testing.With(fields),
+		completed: f.completed,
+	}
+}
+
+func (f *fallbackTestCore) Check(entry zapcore.Entry, checked *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	// see other Check impls, all look similar.
+	// this defers the "where to log" decision to Write, as `f` is the core that will write.
+	if f.fallback.Enabled(entry.Level) {
+		return checked.AddCore(entry, f)
+	}
+	return checked // do not add any cores
+}
+
+func (f *fallbackTestCore) Write(entry zapcore.Entry, fields []zapcore.Field) error {
+	if f.completed.Load() {
+		entry.Message = fmt.Sprintf("COULD FAIL TEST %q, logged too late: %v", f.t.Name(), entry.Message)
+
+		hasStack := slices.ContainsFunc(fields, func(field zapcore.Field) bool {
+			// no specific stack-trace type, so just look for probable fields.
+			return strings.Contains(strings.ToLower(field.Key), "stack")
+		})
+		if !hasStack {
+			fields = append(fields, zap.Stack("log_stack"))
+		}
+		return f.fallback.Write(entry, fields)
+	}
+	// Ensure no concurrent writes to the test logger.
+	f.Lock()
+	defer f.Unlock()
+	return f.testing.Write(entry, fields)
```

Comment:
why the added mutex?  zap cores are race-safe, as is `t.Log` - they flush whole lines without interleaving.

Reply:
From some more reading and thinking and IRL chat: I'm pretty sure this doesn't change anything, but it's not risky or anything.

Now that there's a counter-example showing this logger isn't reliable, I think it's pretty clear it's just that the atomics aren't ensuring "fallback logger is used after test completes".  it currently allows: pausing after checking the atomic is not-completed -> test completes -> resuming and using the `T` logger.

But I can make a PR for that if you want.  The mutex here shouldn't make anything worse, and a mutex will be needed to fix the "atomics are not enough" issue anyway.